### PR TITLE
WriteQuit

### PR DESCRIPTION
**Pull request by srepollock**
### WriteQuit
- `:wq` or `:writequit`
- only works on a file that has already been made and saved already
- need to work on file name parsing for future saving
#198

### CODE REVIEW

File: src/cmd_line/subparser.ts [hunk 1]
```diff
@@ -2,12 +2,16 @@
 
 import {parseQuitCommandArgs} from './subparsers/quit';
 import {parseWriteCommandArgs} from './subparsers/write';
+import {parseWriteQuitCommandArgs} from './subparsers/writequit';
 
 // maps command names to parsers for said commands.
 export const commandParsers = {
     w: parseWriteCommandArgs,
```
Comment (jpoon) [inline src/cmd_line/subparser.ts]: For consistency, can you also stringify 'w' and 'write'?

File: src/cmd_line/commands/writequit.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        this._shortName = "wq";
+        this._arguments = args;
+    }
+
+    get arguments() : IWriteQuitCommandArguments {
+        return this._arguments;
+    }
+
+    // Writing command. Taken as a basis from the "write.ts" file.
+    execute(modeHandler : ModeHandler) : void {
```
Comment (jpoon) [inline src/cmd_line/commands/writequit.ts]: Aside from lines 69, 77, this is the same as what's in `write.ts`. Any way we can just call that function instead? Maybe we can change the `execute` signature to a `bool` to indicate a success and if the write access was successful, go on and do the quit side of things?

File: src/cmd_line/commands/writequit.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            (ok) => {
+                modeHandler.setupStatusBarItem('"' + path.basename(this.activeTextEditor.document.fileName) +
+                '" ' + this.activeTextEditor.document.lineCount + 'L ' +
+                this.activeTextEditor.document.getText().length + 'C written');
+            },
+            (e) => modeHandler.setupStatusBarItem(e)
+        );
+    }
+
+    // Quitting command. Taken as a basis from the "quit.ts" file.
```
Comment (jpoon) [inline src/cmd_line/commands/writequit.ts]: Similar to the above comment, this logic is the same as `quit.ts`

File: src/cmd_line/subparser.ts (hunk 1 above)
Comment (srepollock) [inline src/cmd_line/subparser.ts]: What do you mean by this?

Comment (jpoon) [inline src/cmd_line/subparser.ts]:
```
{
    'w': parseWriteCommandArgs,
    'write': parseWriteCommandArgs,
    ...
}
```

Single-quote around the `w` and `write`

File: src/cmd_line/commands/writequit.ts (hunk 2 above)
Comment (johnfn) [inline src/cmd_line/commands/writequit.ts]: How come we can't just use  `vscode.commands.executeCommand("workbench.action.files.save")`?

Comment (srepollock) [inline src/cmd_line/commands/writequit.ts]: Use this to save the file or to save a new file?

Comment (johnfn) [inline src/cmd_line/commands/writequit.ts]: Save. Maybe there's another one for save as?

Comment (rebornix) [inline src/cmd_line/commands/writequit.ts]: Use `workbench.action.files.saveAs ` command.

Comment (srepollock) [inline src/cmd_line/commands/writequit.ts]: Could that then be used to save the file as a specific name?

Comment (rebornix) [inline src/cmd_line/commands/writequit.ts]: When this command is triggered, a pop up window will show up for user to type file name and location, so we don't need to pass file name to it.

Comment (rebornix) [inline src/cmd_line/commands/writequit.ts]: You can try it by adding below binding to your keybinding configuration.

```
{
        "key": "ctrl+0",
        "command": "workbench.action.files.saveAs",
        "when": ""
    }
```

Comment (srepollock) [inline src/cmd_line/commands/writequit.ts]: Awesome thanks man. I'll take a look tomorrow. Actually just installing Manjaro on my desktop on a different hard drive tonight. Giving i3wm a shot

Comment (johnfn) [inline src/cmd_line/commands/writequit.ts]: 👍

File: src/cmd_line/commands/writequit.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        this._arguments = args;
+    }
+
+    get arguments() : IWriteQuitCommandArguments {
+        return this._arguments;
+    }
+
+    // Writing command. Taken as a basis from the "write.ts" file.
+    execute(modeHandler : ModeHandler) : void {
+        var filename : RegExp = new RegExp("Untitled-[0-9]*");
```
Comment (johnfn) [inline src/cmd_line/commands/writequit.ts]: Hmm, this approach is tightly coupled to how VSCode names untitled documents - if they ever change, then saving will break. Is there any other way to check this?

Comment (srepollock) [inline src/cmd_line/commands/writequit.ts]: I'll take a look

Comment (srepollock) [inline src/cmd_line/commands/writequit.ts]: That was one of the ways I found right through the API. I don't know where else I could get it

Comment (srepollock) [inline src/cmd_line/commands/writequit.ts]: Also it’s how we use it in `src/cmd_line/commands/write.ts` as well

Comment (johnfn) [inline src/cmd_line/commands/writequit.ts]: Eh, if there's really no other way, we can just merge it. It's not that bad.

**Comment by johnfn** [comment]
Finally got this in! Thank you @srepollock!

**Comment by srepollock** [comment]
Thank you guys. This is my first contribution to an open source project that's not my own or one of my friends :smile: hope to keep poking around and help work with this when I can

**Comment by johnfn** [comment]
Congrats! 😄  Hope to see you back again soon.

**Pull request merged**